<div style="font-size: 36px;">
COMS W4111 -- Introduction to Databases, Fall 2025<br>Lecture 6 Examples
</div>

# Initialize Environment

In [1]:
%load_ext sql

In [5]:
db_url = "mysql+pymysql://root:dbuserdbuser@localhost"

In [6]:
# This is a hack to fix a version problem/incompatibility  with some of the packages and magics.
#
%config SqlMagic.style = '_DEPRECATED_DEFAULT' 

In [7]:
%sql $db_url

In [8]:
import pandas

In [9]:
from sqlalchemy import create_engine

In [10]:
engine = create_engine(db_url)

In [11]:
import pymysql

# Functions

## Summary

1. Date and Time
    1. Date and time from string
    2. Quarter, Month, Day of Month, ...
    3. Date diff
    4. Makedate
    5. str_to_date
    6. subtimes
2. IF, IFNULL
3. Type conversation
4. String: Length, Position/Locate/Replace, PAD, Trim

## Date and Time

Manipulating time is common in database applications.

MySQL, like most databases, has powerful [date and time processing functions.](https://dev.mysql.com/doc/refman/8.4/en/date-and-time-functions.html)

### Watching Lectures

In [12]:
stats_file = "session-HeatMap_2025-09-13--2025-10-13_9421bb02-ef10-4150-90e0-b3720120101c.csv"

In [13]:
stats_df = pandas.read_csv(stats_file)
stats_df

,Timestamp,Start Position,Minutes Delivered,UserName,User ID,Name,Email
0,10/10/2025 2:24:21 PM,0.000000,0.0188,CVNCanvas\si2471,9c5f4f3d-e997-4b11-84e2-b1c5015b1c8c,Severin Ihnat,si2471@columbia.edu
1,10/10/2025 2:24:38 PM,0.000000,0.1921,CVNCanvas\si2471,9c5f4f3d-e997-4b11-84e2-b1c5015b1c8c,Severin Ihnat,si2471@columbia.edu
2,10/10/2025 2:24:38 PM,11.556562,0.2125,CVNCanvas\si2471,9c5f4f3d-e997-4b11-84e2-b1c5015b1c8c,Severin Ihnat,si2471@columbia.edu
3,10/10/2025 2:25:48 PM,11.556562,2.1763,CVNCanvas\si2471,9c5f4f3d-e997-4b11-84e2-b1c5015b1c8c,Severin Ihnat,si2471@columbia.edu
4,10/10/2025 2:27:34 PM,213.363729,3.5425,CVNCanvas\si2471,9c5f4f3d-e997-4b11-84e2-b1c5015b1c8c,Severin Ihnat,si2471@columbia.edu
...,...,...,...,...,...,...,...
491,10/12/2025 10:17:18 AM,236.600168,1.7233,CVNCanvas\dav2136,bc1702bd-70a9-4d25-a102-b1a1010e25ac,Dominick Vaske,dav2136@columbia.edu
492,10/12/2025 10:17:48 AM,359.701977,0.0232,CVNCanvas\dav2136,bc1702bd-70a9-4d25-a102-b1a1010e25ac,Dominick Vaske,dav2136@columbia.edu
493,10/12/2025 10:32:48 AM,370.767161,30.3812,CVNCanvas\dav2136,bc1702bd-70a9-4d25-a102-b1a1010e25ac,Dominick Vaske,dav2136@columbia.edu
494,10/12/2025 10:55:48 AM,2063.400047,46.2985,CVNCanvas\dav2136,bc1702bd-70a9-4d25-a102-b1a1010e25ac,Dominick Vaske,dav2136@columbia.edu


In [15]:
%sql create schema if not exists f2025_w4111_examples

 * mysql+pymysql://root:***@localhost
1 rows affected.


[]

In [16]:
stats_df.to_sql(
    "Heatmap_2025_10_13",
    schema="f2025_w4111_examples",
    index=False,
    if_exists="replace",
    con=engine
)

496

In [17]:
%sql use f2025_w4111_examples

 * mysql+pymysql://root:***@localhost
0 rows affected.


[]

In [18]:
%sql describe Heatmap_2025_10_13

 * mysql+pymysql://root:***@localhost
7 rows affected.


Field,Type,Null,Key,Default,Extra
Timestamp,text,YES,,None,
Start Position,double,YES,,None,
Minutes Delivered,double,YES,,None,
UserName,text,YES,,None,
User ID,text,YES,,None,
Name,text,YES,,None,
Email,text,YES,,None,


In [19]:
%sql select * from heatmap_2025_10_13 limit 5

 * mysql+pymysql://root:***@localhost
5 rows affected.


Timestamp,Start Position,Minutes Delivered,UserName,User ID,Name,Email
10/10/2025 2:24:21 PM,0.0,0.0188,CVNCanvas\si2471,9c5f4f3d-e997-4b11-84e2-b1c5015b1c8c,Severin Ihnat,si2471@columbia.edu
10/10/2025 2:24:38 PM,0.0,0.1921,CVNCanvas\si2471,9c5f4f3d-e997-4b11-84e2-b1c5015b1c8c,Severin Ihnat,si2471@columbia.edu
10/10/2025 2:24:38 PM,11.556562,0.2125,CVNCanvas\si2471,9c5f4f3d-e997-4b11-84e2-b1c5015b1c8c,Severin Ihnat,si2471@columbia.edu
10/10/2025 2:25:48 PM,11.556562,2.1763,CVNCanvas\si2471,9c5f4f3d-e997-4b11-84e2-b1c5015b1c8c,Severin Ihnat,si2471@columbia.edu
10/10/2025 2:27:34 PM,213.363729,3.5425,CVNCanvas\si2471,9c5f4f3d-e997-4b11-84e2-b1c5015b1c8c,Severin Ihnat,si2471@columbia.edu


This is in the format:
%m/%d/%Y %h:%i:%s %p

Where:

- %m = month (2-digit)

- %d = day (2-digit)

- %Y = 4-digit year

- %h = hour (1-12)

- %i = minutes

- %s = seconds

- %p = AM or PM

In [20]:
%%sql

SELECT STR_TO_DATE('10/10/2025 2:24:21 PM', '%m/%d/%Y %h:%i:%s %p') AS view_datetime,
       `Start Position`                                             as start_position_seconds,
       sec_to_time(round(Heatmap_2025_10_13.`Start Position`, 0))   as start_position_hms,
       `Minutes Delivered`                                          as minutes_delivered,
       sec_to_time(round(`Minutes Delivered` * 60, 0))              as minutes_delivered_hms,
       substr(UserName, position('\\' in UserName) + 1)             as user_name
from Heatmap_2025_10_13
limit 10;


 * mysql+pymysql://root:***@localhost
10 rows affected.


view_datetime,start_position_seconds,start_position_hms,minutes_delivered,minutes_delivered_hms,user_name
2025-10-10 14:24:21,0.0,0:00:00,0.0188,0:00:01,si2471
2025-10-10 14:24:21,0.0,0:00:00,0.1921,0:00:12,si2471
2025-10-10 14:24:21,11.556562,0:00:12,0.2125,0:00:13,si2471
2025-10-10 14:24:21,11.556562,0:00:12,2.1763,0:02:11,si2471
2025-10-10 14:24:21,213.363729,0:03:33,3.5425,0:03:33,si2471
2025-10-10 14:24:21,466.405459,0:07:46,0.0174,0:00:01,si2471
2025-10-10 14:24:21,5437.900185,1:30:38,0.0375,0:00:02,nk3214
2025-10-10 14:24:21,5360.700481,1:29:21,0.4029,0:00:24,nk3214
2025-10-10 14:24:21,5394.847015,1:29:55,0.2865,0:00:17,nk3214
2025-10-10 14:24:21,0.0,0:00:00,0.013,0:00:01,dff9


In [24]:
%%sql

with one as (
    SELECT STR_TO_DATE('10/10/2025 2:24:21 PM', '%m/%d/%Y %h:%i:%s %p') AS view_datetime,
       `Start Position`                                             as start_position_seconds,
       sec_to_time(round(Heatmap_2025_10_13.`Start Position`, 0))   as start_position_hms,
       `Minutes Delivered`                                          as minutes_delivered,
       sec_to_time(round(`Minutes Delivered` * 60, 0))              as minutes_delivered_hms,
       substr(UserName, position('\\' in UserName) + 1)             as user_name
from Heatmap_2025_10_13
),
two as (
    select
    sum(minutes_delivered) as total_minutes,
    user_name
from
    one
group by user_name)
select
    user_name,
    total_minutes,
    sec_to_time(total_minutes*60) as total_hms
from two
order by total_minutes desc;

 * mysql+pymysql://root:***@localhost
10 rows affected.


user_name,total_minutes,total_hms
shl2183,133.54059999999998,2:13:32.436000
dav2136,118.78459999999998,1:58:47.076000
yy3608,116.38910000000001,1:56:23.346000
si2471,115.1609,1:55:09.654000
nk3214,90.9545,1:30:57.270000
fm2859,47.8448,0:47:50.688000
mh4605,45.82959999999999,0:45:49.776000
cl4537,31.2095,0:31:12.570000
ss7544,1.0477,0:01:02.862000
dff9,0.3378,0:00:20.268000


### Sales and Orders Analysis Breakdown

| <img src="classic_models.jpg" width=700px> |
| :---: |
| __Classic Model Schema__ |


#### How bad are the shipping delays?

In [25]:
%%sql
use classicmodels;

with one as (
    select
        orderNumber,
        shippedDate,
        orderDate,
        datediff(shippedDate, orderDate) as shipping_days,
        (select customerName from customers where
                                                customers.customerNumber=orders.customerNumber) as customerName
        from orders
        where shippedDate is not null
)
select
    *
from
    one
order by shipping_days desc;

 * mysql+pymysql://root:***@localhost
0 rows affected.
312 rows affected.


orderNumber,shippedDate,orderDate,shipping_days,customerName
10165,2003-12-26,2003-10-22,65,"Dragon Souveniers, Ltd."
10210,2004-01-20,2004-01-12,8,Osaka Souveniers Co.
10258,2004-06-23,2004-06-15,8,"Tokyo Collectables, Ltd"
10240,2004-04-20,2004-04-13,7,Osaka Souveniers Co.
10339,2004-11-30,2004-11-23,7,"Tokyo Collectables, Ltd"
10121,2003-05-13,2003-05-07,6,Reims Collectables
10133,2003-07-03,2003-06-27,6,Euro+ Shopping Channel
10138,2003-07-13,2003-07-07,6,Kelly's Gift Shop
10140,2003-07-30,2003-07-24,6,Technics Stores Inc.
10145,2003-08-31,2003-08-25,6,Toys4GrownUps.com


Lets look at summaries for customers.

In [26]:
%%sql

with one as (
    select
        orderNumber,
        shippedDate,
        orderDate,
        datediff(shippedDate, orderDate) as shipping_days,
        (select customerName from customers where
                                                customers.customerNumber=orders.customerNumber) as customerName
        from orders
        where shippedDate is not null
)
select
    customerName,
    count(*) as total_orders,
    min(shipping_days) as min_shipping_days,
    max(shipping_days) as max_shipping_days,
    avg(shipping_days) as avg_shipping_days
from
    one
group by customerName
order by avg_shipping_days desc;

 * mysql+pymysql://root:***@localhost
98 rows affected.


customerName,total_orders,min_shipping_days,max_shipping_days,avg_shipping_days
"Dragon Souveniers, Ltd.",5,1,65,14.6000
Osaka Souveniers Co.,2,7,8,7.5000
Mini Caravy,3,5,6,5.6667
Auto-Moto Classics Inc.,3,5,6,5.6667
"Tokyo Collectables, Ltd",4,2,8,5.5000
Scandinavian Gift Ideas,2,5,6,5.5000
CAF Imports,2,5,6,5.5000
Boards & Toys Co.,2,5,6,5.5000
giftsbymail.co.uk,2,5,6,5.5000
Cruz & Sons Co.,3,5,6,5.3333


#### Sales by Year, Quarter, Month

In [27]:
%%sql

with one as (
    select
        orderNumber, quantityOrdered, priceEach, quantityOrdered*priceEach as revenue,
        shippedDate,
        year(shippedDate) as shipYear,
        quarter(shippedDate) as shipQuarter,
        month(shippedDate) as shipMonth
    from orders join orderdetails using(orderNumber)
    where shippedDate is not null
)
select
    shipYear, shipQuarter, shipMonth,
    concat("$", format(sum(revenue), 2)) as total_revenue
from
    one
group by shipYear, shipQuarter, shipMonth
order by shipYear, shipQuarter, shipMonth;


 * mysql+pymysql://root:***@localhost
29 rows affected.


shipYear,shipQuarter,shipMonth,total_revenue
2003,1,1,"$26,267.62"
2003,1,2,"$218,828.79"
2003,1,3,"$160,517.14"
2003,2,4,"$104,158.23"
2003,2,5,"$203,993.99"
2003,2,6,"$185,236.65"
2003,3,7,"$224,306.40"
2003,3,8,"$178,257.11"
2003,3,9,"$181,936.68"
2003,4,10,"$406,793.79"


__I have had about as much fun with dates and times as I can stand for one day.__

## IF, IFNULL

Consider revenue per customer for Classic Models.

In [28]:
%%sql

/*
    How many customers are there?
*/
select count(*) as no_of_customers from classicmodels.customers;

 * mysql+pymysql://root:***@localhost
1 rows affected.


no_of_customers
122


In [29]:
%%sql

with one as (
    select customerNumber, orderNumber, quantityOrdered*priceEach as revenue from orders join orderdetails using(orderNumber)
),
    two as (
        select customerNumber, sum(revenue) as total_revenue from one group by customerNumber
    )
select
    customerNumber,
    (select customerName from customers where customers.customerNumber=two.customerNumber) as customerName,
    total_revenue from two;

 * mysql+pymysql://root:***@localhost
98 rows affected.


customerNumber,customerName,total_revenue
103,Atelier graphique,22314.36
112,Signal Gift Stores,80180.98
114,"Australian Collectors, Co.",180585.07
119,La Rochelle Gifts,158573.12
121,Baane Mini Imports,104224.79
124,Mini Gifts Distributors Ltd.,591827.34
128,"Blauer See Auto, Co.",75937.76
129,Mini Wheels Co.,66710.56
131,Land of Toys Inc.,149085.15
141,Euro+ Shopping Channel,820689.54


I have 98 entries but 122 customers. Clearly some are missing. Let's try a slightly different approach

In [30]:
%%sql

with one as (
    select customerNumber, orderNumber, quantityOrdered*priceEach as revenue from orders join orderdetails using(orderNumber)
),
    two as (
        select customerNumber, sum(revenue) as total_revenue from one group by customerNumber
    )
select
    customerNumber, customerName, total_revenue from customers left join two using(customerNumber)
    order by total_revenue;

 * mysql+pymysql://root:***@localhost
122 rows affected.


customerNumber,customerName,total_revenue
125,Havel & Zbyszek Co,None
168,American Souvenirs Inc,None
169,Porto Imports Co.,None
206,"Asian Shopping Network, Co",None
223,Natürlich Autos,None
237,ANG Resellers,None
247,Messner Shopping Network,None
273,"Franken Gifts, Co",None
293,BG&E Collectables,None
303,Schuyler Imports,None


Well, ```NULL``` maps to ```None``` in Python, but we know the answer. It is ```0```.

In [31]:
%%sql

with one as (
    select customerNumber, orderNumber, quantityOrdered*priceEach as revenue from orders join orderdetails using(orderNumber)
),
    two as (
        select customerNumber, sum(revenue) as total_revenue from one group by customerNumber
    )
select
    customerNumber, customerName, ifnull(total_revenue,0) as total_revenue from customers left join two using(customerNumber)
    order by total_revenue;

 * mysql+pymysql://root:***@localhost
122 rows affected.


customerNumber,customerName,total_revenue
125,Havel & Zbyszek Co,0.00
168,American Souvenirs Inc,0.00
169,Porto Imports Co.,0.00
206,"Asian Shopping Network, Co",0.00
223,Natürlich Autos,0.00
237,ANG Resellers,0.00
247,Messner Shopping Network,0.00
273,"Franken Gifts, Co",0.00
293,BG&E Collectables,0.00
303,Schuyler Imports,0.00


This example of ```IF``` is a little contrived, but it does show the behavior.

Also, note that I had to use a subquery to get the average. I could not just call ```avg```.

In [32]:
%%sql

with one as (
    select customerNumber, orderNumber, quantityOrdered*priceEach as revenue from orders join orderdetails using(orderNumber)
),
    two as (
        select customerNumber, sum(revenue) as total_revenue from one group by customerNumber
    ),
three as (
    select
    customerNumber, customerName, ifnull(total_revenue, 0) as total_revenue from customers left join two using (customerNumber)
    order by total_revenue
    )
select
    customerNumber, customerName, total_revenue,
        if(total_revenue >= (select avg(total_revenue) from three), "Good Customer", "Bad Customer") customer_kind
from
    three;

 * mysql+pymysql://root:***@localhost
122 rows affected.


customerNumber,customerName,total_revenue,customer_kind
125,Havel & Zbyszek Co,0.00,Bad Customer
168,American Souvenirs Inc,0.00,Bad Customer
169,Porto Imports Co.,0.00,Bad Customer
206,"Asian Shopping Network, Co",0.00,Bad Customer
223,Natürlich Autos,0.00,Bad Customer
237,ANG Resellers,0.00,Bad Customer
247,Messner Shopping Network,0.00,Bad Customer
273,"Franken Gifts, Co",0.00,Bad Customer
293,BG&E Collectables,0.00,Bad Customer
303,Schuyler Imports,0.00,Bad Customer


## Type Conversion

There are two basic, nearly the same, functions:
1. ```CAST()```
2. ```CONVERT()```

I have found them do be useful mostly when dealing with importing data from files, e.g. CSV, JSON.

In [35]:
players = [
    {
        "playerID": "gehrilo01",
        "nameFirst": "Lou",
        "nameLast": "Gehrig",
        "birthYear": 1903,
        "birthCountry": "USA",
        "bats": "L",
        "throws": "L",
        "weight": 200.0
    },
    {
        "playerID": "willite01",
        "nameFirst": "Ted",
        "nameLast": "Williams",
        "birthYear": 1918,
        "birthCountry": "USA",
        "bats": "L",
        "throws": "R",
        "weight": 205.0
    },
    {
        "playerID": "mantlmi01",
        "nameFirst": "Mickey",
        "nameLast": "Mantle",
        "birthYear": 1931,
        "birthCountry": "USA",
        "bats": "B",
        "throws": "R",
        "weight": 195.0
    },
    {
        "playerID": "robinja02",
        "nameFirst": "Jackie",
        "nameLast": "Robinson",
        "birthYear": 1919,
        "birthCountry": "USA",
        "bats": "R",
        "throws": "R",
        "weight": 195.0
    },
    {
        "playerID": "dimagjo01",
        "nameFirst": "Joe",
        "nameLast": "DiMaggio",
        "birthYear": 1914,
        "birthCountry": "USA",
        "bats": "R",
        "throws": "R",
        "weight": 193.0
    },
    {
        "playerID": "musialst01",
        "nameFirst": "Stan",
        "nameLast": "Musial",
        "birthYear": 1920,
        "birthCountry": "USA",
        "bats": "L",
        "throws": "L",
        "weight": 175.0
    },
    {
        "playerID": "koufasa01",
        "nameFirst": "Sandy",
        "nameLast": "Koufax",
        "birthYear": 1935,
        "birthCountry": "USA",
        "bats": "L",
        "throws": "L",
        "weight": 210.0
    },
    {
        "playerID": "jeterde01",
        "nameFirst": "Derek",
        "nameLast": "Jeter",
        "birthYear": 1974,
        "birthCountry": "USA",
        "bats": "R",
        "throws": "R",
        "weight": 195.0
    },
    {
        "playerID": "pujolal01",
        "nameFirst": "Albert",
        "nameLast": "Pujols",
        "birthYear": 1980,
        "birthCountry": "Dominican Republic",
        "bats": "R",
        "throws": "R",
        "weight": 230.0
    },
    {
        "playerID": "troutmi01",
        "nameFirst": "Mike",
        "nameLast": "Trout",
        "birthYear": 1991,
        "birthCountry": "USA",
        "bats": "R",
        "throws": "R",
        "weight": 235.0
    },
    {
        "playerID": "ohtansh01",
        "nameFirst": "Shohei",
        "nameLast": "Ohtani",
        "birthYear": 1994,
        "birthCountry": "Japan",
        "bats": "L",
        "throws": "R",
        "weight": 210.0
    }
]



In [36]:
df = pandas.DataFrame(players)
df

,playerID,nameFirst,nameLast,birthYear,birthCountry,bats,throws,weight
0,gehrilo01,Lou,Gehrig,1903,USA,L,L,200.0
1,willite01,Ted,Williams,1918,USA,L,R,205.0
2,mantlmi01,Mickey,Mantle,1931,USA,B,R,195.0
3,robinja02,Jackie,Robinson,1919,USA,R,R,195.0
4,dimagjo01,Joe,DiMaggio,1914,USA,R,R,193.0
5,musialst01,Stan,Musial,1920,USA,L,L,175.0
6,koufasa01,Sandy,Koufax,1935,USA,L,L,210.0
7,jeterde01,Derek,Jeter,1974,USA,R,R,195.0
8,pujolal01,Albert,Pujols,1980,Dominican Republic,R,R,230.0
9,troutmi01,Mike,Trout,1991,USA,R,R,235.0


In [37]:
df.to_sql(
    "Simple_Players",
    schema="f2025_w4111_examples",
    index=False,
    if_exists="replace",
    con=engine
)

11

In [38]:
%sql use f2025_w4111_examples;

 * mysql+pymysql://root:***@localhost
0 rows affected.


[]

In [39]:
%sql describe Simple_players;

 * mysql+pymysql://root:***@localhost
8 rows affected.


Field,Type,Null,Key,Default,Extra
playerID,text,YES,,None,
nameFirst,text,YES,,None,
nameLast,text,YES,,None,
birthYear,bigint,YES,,None,
birthCountry,text,YES,,None,
bats,text,YES,,None,
throws,text,YES,,None,
weight,double,YES,,None,


Pandas and other import tools "guess" the data types from the values in the file. The tools often gues wrong. Guessing that
- An int column is double.
- A year is an int
  
are common examples. There are others.

In [40]:
%%sql

drop table if exists Simple_Players_Fixed;

create table Simple_Players_Fixed as 
select
    playerID, nameFirst, nameLast,
    cast(birthYear as Year) as birthYear,
    birthCountry,
    bats, throws,
    cast(weight as signed int) as weight
from
    Simple_players;

 * mysql+pymysql://root:***@localhost
0 rows affected.
11 rows affected.


[]

In [41]:
%sql describe Simple_Players_Fixed;

 * mysql+pymysql://root:***@localhost
8 rows affected.


Field,Type,Null,Key,Default,Extra
playerID,text,YES,,None,
nameFirst,text,YES,,None,
nameLast,text,YES,,None,
birthYear,year,YES,,None,
birthCountry,text,YES,,None,
bats,text,YES,,None,
throws,text,YES,,None,
weight,bigint,YES,,None,


Note that in a lot of cases, e.g. INSERT into a table, the database engine tries to automatical CAST/CONVERT the type.

In [42]:
%%sql

select playerID,
        birthYear+12.7 as funnyNumber,
        concat(weight, " is how much the players weighs") as funnyWeight,
        concat(birthYear, " plus ", weight, " plus 12.7 is ", birthYear+weight+12.7) as reallyWeird
from Simple_Players_Fixed;

 * mysql+pymysql://root:***@localhost
11 rows affected.


playerID,funnyNumber,funnyWeight,reallyWeird
gehrilo01,1915.7,200 is how much the players weighs,1903 plus 200 plus 12.7 is 2115.7
willite01,1930.7,205 is how much the players weighs,1918 plus 205 plus 12.7 is 2135.7
mantlmi01,1943.7,195 is how much the players weighs,1931 plus 195 plus 12.7 is 2138.7
robinja02,1931.7,195 is how much the players weighs,1919 plus 195 plus 12.7 is 2126.7
dimagjo01,1926.7,193 is how much the players weighs,1914 plus 193 plus 12.7 is 2119.7
musialst01,1932.7,175 is how much the players weighs,1920 plus 175 plus 12.7 is 2107.7
koufasa01,1947.7,210 is how much the players weighs,1935 plus 210 plus 12.7 is 2157.7
jeterde01,1986.7,195 is how much the players weighs,1974 plus 195 plus 12.7 is 2181.7
pujolal01,1992.7,230 is how much the players weighs,1980 plus 230 plus 12.7 is 2222.7
troutmi01,2003.7,235 is how much the players weighs,1991 plus 235 plus 12.7 is 2238.7


# Some More Fun with Strings

In [43]:
title_basics_df = pandas.read_csv("got_imdb_title_basics.csv")

In [44]:
title_basics_df.head(10)

,tconst,title_type,primary_title,alternate_title,is_adult,start_year,end_year,run_time_minutes,genres
0,tt0054518,tvSeries,The Avengers,The Avengers,0,1961.0,1969.0,50.0,"Action,Comedy,Crime"
1,tt0054571,tvSeries,Three Live Wires,Three Live Wires,0,1961.0,NaN,30.0,Comedy
2,tt0055556,movie,"Two Living, One Dead","Two Living, One Dead",0,1961.0,NaN,105.0,"Crime,Drama,Thriller"
3,tt0056105,movie,The Swingin' Maiden,The Iron Maiden,0,1962.0,NaN,98.0,"Comedy,Romance"
4,tt0056696,movie,Young and Willing,The Wild and the Willing,0,1962.0,NaN,110.0,"Drama,Romance"
5,tt0057435,movie,The Punch and Judy Man,The Punch and Judy Man,0,1963.0,NaN,96.0,Comedy
6,tt0058142,movie,Girl with Green Eyes,Girl with Green Eyes,0,1964.0,NaN,91.0,"Drama,Romance"
7,tt0058596,movie,Smokescreen,Smokescreen,0,1964.0,NaN,70.0,"Crime,Drama,Mystery"
8,tt0059106,movie,Die! Die! My Darling!,Fanatic,0,1965.0,NaN,97.0,"Horror,Thriller"
9,tt0059191,tvEpisode,For the West,For the West,0,1965.0,NaN,75.0,Drama


In [45]:
title_basics_df.to_sql(
    "title_basics_raw", schema="f2025_w4111_examples", con=engine, index=False, if_exists="replace"
)

29058

In [46]:
%sql select * from title_basics_raw limit 10

 * mysql+pymysql://root:***@localhost
10 rows affected.


tconst,title_type,primary_title,alternate_title,is_adult,start_year,end_year,run_time_minutes,genres
tt0054518,tvSeries,The Avengers,The Avengers,0,1961.0,1969.0,50.0,"Action,Comedy,Crime"
tt0054571,tvSeries,Three Live Wires,Three Live Wires,0,1961.0,None,30.0,Comedy
tt0055556,movie,"Two Living, One Dead","Two Living, One Dead",0,1961.0,None,105.0,"Crime,Drama,Thriller"
tt0056105,movie,The Swingin' Maiden,The Iron Maiden,0,1962.0,None,98.0,"Comedy,Romance"
tt0056696,movie,Young and Willing,The Wild and the Willing,0,1962.0,None,110.0,"Drama,Romance"
tt0057435,movie,The Punch and Judy Man,The Punch and Judy Man,0,1963.0,None,96.0,Comedy
tt0058142,movie,Girl with Green Eyes,Girl with Green Eyes,0,1964.0,None,91.0,"Drama,Romance"
tt0058596,movie,Smokescreen,Smokescreen,0,1964.0,None,70.0,"Crime,Drama,Mystery"
tt0059106,movie,Die! Die! My Darling!,Fanatic,0,1965.0,None,97.0,"Horror,Thriller"
tt0059191,tvEpisode,For the West,For the West,0,1965.0,None,75.0,Drama


In [ ]:
%sql describe title_basics_raw;

You can see that ```genres``` is multi-valued, and the individual values domain seems to be a fixed set of values, i.e. an ```ENUM.```

You can also infer that a ```title_basics``` has many ```genres``` and a ```genre``` may apply to several ```title_basics.``` $\Rightarrow$ We need an associative entity.

A better model is 

| <img src="multi-valued-erd.jpg"> |
| :---: |
| __Multi-Valued Attributes__ |



It would be relatively easy to write some python code that queries the table, splits the genre into a separate dataframe, creates a dataframe with the associations and write all of the entities to the database.

This approach works for relatively small datasets.

If the dataset is very large, this can be inefficient/expensive. The notebook and database are different programs running in different processes, perhaps on different machines $\Rightarrow$ That is a lot of data movement.

We will also see that advanced databases can parallelize and perform other optimizations on computationally complex queries.

Let's do some work to transform _in the database_ without moving the data.

In [47]:
%%sql

select genres,
            length(genres),
            replace(genres, ',', '') as no_commas,
            length(replace(genres, ',','')) as length_no_commas,
            length(genres) - length(replace(genres, ',','')) + 1 as no_of_genres
    from
        title_basics_raw
order by length(genres) - length(replace(genres, ',','')) + 1 desc
limit 5;

 * mysql+pymysql://root:***@localhost
5 rows affected.


genres,length(genres),no_commas,length_no_commas,no_of_genres
"Drama,Fantasy,Horror",20,DramaFantasyHorror,18,3
"Crime,Drama,Mystery",19,CrimeDramaMystery,17,3
"Comedy,Crime,Drama",18,ComedyCrimeDrama,16,3
"Horror,Mystery,Thriller",23,HorrorMysteryThriller,21,3
"Horror,Mystery,Thriller",23,HorrorMysteryThriller,21,3


This query "parses" and "splits" the genres.

In [48]:
%%sql

with one as (
    select
        genres,
        if(substr(genres, 1, locate(',', genres)-1)='', null, substr(genres, 1, locate(',', genres)-1)) as g1,
        if(substr(genres, locate(',', genres)+1)='', null, substr(genres, locate(',', genres)+1)) as remainder
    from
        title_basics_raw
),
two as (
    select
        genres, g1,
        if(substr(remainder, 1, locate(',', remainder)-1)='', null, substr(remainder, 1, locate(',', remainder)-1)) as g2,
        if(substr(remainder, locate(',', genres)+1)='', null, substr(remainder, locate(',', remainder)+1)) as g3
    from one
    )
select * from two;

 * mysql+pymysql://root:***@localhost
29058 rows affected.


genres,g1,g2,g3
"Action,Comedy,Crime",Action,Comedy,Crime
Comedy,None,None,Comedy
"Crime,Drama,Thriller",Crime,Drama,Thriller
"Comedy,Romance",Comedy,None,None
"Drama,Romance",Drama,None,Romance
Comedy,None,None,Comedy
"Drama,Romance",Drama,None,Romance
"Crime,Drama,Mystery",Crime,Drama,Mystery
"Horror,Thriller",Horror,None,Thriller
Drama,None,None,Drama


With a little bit more work, I can
- Get the unique values for g1, g2, g3.
- Make a table with the unique values and create an auto-increment primary key.
- Build the associative entity link the genres with the titles.
- Delete the genres column from titles.

This is an example of the __fun data engineering__ that students will perform on the non-programming homeworks/project.

Have fun.

| <img src="fun.jpg" width-=500px> |
| :---: |
| __Fun__ |

# Index Example

[Lahman's Baseball Database](https://www.kaggle.com/datasets/freshrenzo/lahmanbaseballdatabase) is an interesting dataset for data engineering and analysis.

In [ ]:
people_df = pandas.read_csv("./people.csv")

In [ ]:
batting_df = pandas.read_csv("./batting.csv")

In [ ]:
appearances_df = pandas.read_csv("./appearances.csv")

In [ ]:
%sql drop table if exists people;
%sql drop table if exists batting;
%sql drop table if exists appearances;

In [ ]:
%%sql

drop index  batting_playerID_index on batting;

drop index  people_player_ID_index on people;


In [ ]:
people_df.to_sql(
    "people",
    schema="f2025_w4111_examples",
    con=engine,
    index=False,
    if_exists="replace"
)

In [ ]:
batting_df.to_sql(
    "batting",
    schema="f2025_w4111_examples",
    con=engine,
    index=False,
    if_exists="replace"
)

In [ ]:
appearances_df.to_sql(
    "appearances",
    schema="f2025_w4111_examples",
    con=engine,
    index=False,
    if_exists="replace"
)

In [ ]:
from time import time

start_time = time()

for i in range(0, 1):
    result = %sql with one as (select playerID, \
                        nameLast, \
                        nameFirst, \
                        sum(H)   as H, \
                        sum(AB)  as AB, \
                        sum(BB)  as BB, \
                        sum(HR)  as HR, \
                        sum(RBI) as RBI, \
                        ( \
                            (sum(H) + sum(BB)) / \
                            (if(sum(AB) + sum(BB) = 0, NULL, sum(AB) + sum(BB))) \
                            ) as obp \
                 from people \
                          join batting using (playerID) \
                 group by playerID, nameLast, nameFirst) \
        select * from one  where playerID='willite01' and AB > 1000
    
end_time = time()

elapsed_time = (end_time - start_time)

print ("Elapsed time = ", elapsed_time)

In [ ]:
from time import time

start_time = time()

for i in range(0, 1):
    result = %sql select * from people where nameLast="Williams"
    
end_time = time()

elapsed_time = (end_time - start_time)

print ("Elapsed time = ", elapsed_time)

In [ ]:
%%sql

alter table batting
    modify playerID varchar(16) null;

alter table people
    modify playerID varchar(16) null;

alter table people
    modify nameLast varchar(128) null;

create index batting_playerID_index
    on batting (playerID);

create index people_player_ID_index
    on people(playerID);

In [ ]:
from time import time

start_time = time()

for i in range(0, 1):
    result = %sql select * from people where nameLast="Williams"
    
end_time = time()

elapsed_time = (end_time - start_time)

print ("Elapsed time = ", elapsed_time)

In [ ]:
from time import time

start_time = time()

for i in range(0, 1):
    result = %sql with one as (select playerID, \
                        nameLast, \
                        nameFirst, \
                        sum(H)   as H, \
                        sum(AB)  as AB, \
                        sum(BB)  as BB, \
                        sum(HR)  as HR, \
                        sum(RBI) as RBI, \
                        ( \
                            (sum(H) + sum(BB)) / \
                            (if(sum(AB) + sum(BB) = 0, NULL, sum(AB) + sum(BB))) \
                            ) as obp \
                 from people \
                          join batting using (playerID) \
                 group by playerID, nameLast, nameFirst) \
        select * from one  where playerID='willite01';
    
end_time = time()

elapsed_time2 = (end_time - start_time)

print ("Elapsed time = ", elapsed_time)

# Worked Example: Student, Faculty, Person

## Base Tables

In [ ]:
%%sql 

use f2025_w4111_examples;

drop table if exists Person;

create table Person
(
    UNI             varchar(12)  not null,
    last_name       varchar(64)  not null,
    first_name      varchar(64)  not null,
    preferred_email varchar(128) not null,
    /*
     Since I am putting an index on default email, it must be stored.
     Since the UNI is unique, I do not need to worry about a unique constraint.
     */
    default_email   varchar(128) as (concat(UNI, '@columbia.edu')) stored,
    constraint Person_pk
        primary key (UNI),
    constraint Person_pk_2
        unique (preferred_email)
);

create index Person_default_email_index
    on Person (default_email);

In [ ]:
%%sql

drop table if exists Faculty;

create table Faculty
(
    UNI        varchar(12)                                                                                                          not null,
    department varchar(4)                                                                                                           not null,
    title      enum ('Professor', 'Associate Professor', 'Assistant Professor', 'Adjunct Professor', 'Lecturer', 'Senior Lecturer') not null,
    salary     double                                                                                                               not null,
    constraint Faculty_pk
        primary key (UNI),
    constraint Faculty_Person_UNI_fk
        foreign key (UNI) references Person (UNI)
);

In [ ]:
%%sql

drop table if exists Student;

create table Student
(
    UNI             varchar(12)                 not null,
    graduation_year varchar(4)                  not null,
    major           varchar(32)                 not null,
    constraint Student_pk
        primary key (UNI),
    constraint Student_Person_UNI_fk
        foreign key (UNI) references Person (UNI)
);

## Views

In [ ]:
%%sql

create or replace view FacultyPublic as
    select
        Person.*,
        department, title
    from
        Person JOIN Faculty using(UNI);

create or replace view StudentPublic as
    select
        *
    from
        Person NATURAL JOIN Student;

## Computing UNI

In [ ]:
%%sql

drop function if exists compute_next_uni;


CREATE FUNCTION compute_next_uni(first_name varchar(64), last_name varchar(64))
    RETURNS varchar(12)
    DETERMINISTIC
BEGIN
    -- Declare local variables if needed
    DECLARE result varchar(12);
    declare first_initial char(1);
    declare last_initial char(1);
    declare uni_prefix char(3);
    declare prefix_count int;

    set first_initial = lower(substr(first_name, 1, 1));
    set last_initial = lower(substr(last_name, 1, 1));
    set uni_prefix = concat(first_initial, last_initial, '%');

    set prefix_count = (select count(*) from person where uni like uni_prefix);

    set result = concat(first_initial, last_initial, prefix_count + 1);

    RETURN result;
END;



## Procedures

In [ ]:
%%sql

drop procedure if exists create_faculty;

CREATE DEFINER=`root`@`localhost` PROCEDURE create_faculty (
    IN new_first_name varchar(64),
    IN new_last_name varchar(64),
    IN new_preferred_email varchar(128),
    in new_title enum('Professor', 'Associate Professor', 'Assistant Professor', 'Adjunct Professor', 'Lecturer', 'Senior Lecturer'),
    in new_salary double,
    in new_department varchar(4),
    OUT new_uni varchar(12)
)
SQL SECURITY DEFINER
BEGIN
    set new_uni = compute_next_uni(new_first_name, new_last_name);
    insert into Person(UNI, first_name, last_name, preferred_email)
        value(new_uni, new_first_name, new_last_name, new_preferred_email);
    insert into Faculty(UNI,department, title, salary)
        value(new_uni, new_department, new_title, new_salary);

END;

drop procedure if exists create_student;

CREATE DEFINER=`root`@`localhost` PROCEDURE create_student (
    IN new_first_name varchar(64),
    IN new_last_name varchar(64),
    IN new_preferred_email varchar(128),
    IN new_graduation_year YEAR,
    IN new_major varchar(32),
    OUT new_uni varchar(12)
)
SQL SECURITY DEFINER
BEGIN
    set new_uni = compute_next_uni(new_first_name, new_last_name);
    insert into Person(UNI, first_name, last_name, preferred_email)
        value(new_uni, new_first_name, new_last_name, new_preferred_email);
    insert into Student(UNI, graduation_year, major)
        value(new_uni, new_graduation_year, new_major);

END;

## Some Constraints

Let's put some constraints:
1. ```preferred_email``` must contain ```@```.
2. ```salary``` must be greater than ```0```.

In [ ]:
%%sql

alter table Person
    add constraint check_preffered_email
        check (LOCATE('@', Person.preferred_email) > 0);

alter table Faculty
    add constraint check_salary
        check (salary >= 0);

## Some Security

General users should not be able to mess around with ```Faculty``` and see the ```salary```.

In [ ]:
%%sql

create user general_user
    identified by 'dbuserdbuser';

In [ ]:
%sql mysql+pymysql://general_user:dbuserdbuser@localhost --alias general_user_connection

In [ ]:
%sql general_user@localhost select * from f2025_w4111_examples.Faculty;

The default is that a user has not GRANTs or priveliges.

Let's allow the user to query ```Person```.

In [ ]:
%sql root@localhost

%sql GRANT SELECT ON f2025_w4111_examples.Person TO 'general_user' @'%'

In [ ]:
%sql general_user@localhost select * from f2025_w4111_examples.Person

In [ ]:
#
# Let's add some additional grants.
#
%sql root@localhost GRANT SELECT ON f2025_w4111_examples.FacultyPublic TO 'general_user' @'%';
%sql root@localhost GRANT SELECT ON f2025_w4111_examples.StudentPublic TO 'general_user' @'%';

In [ ]:
%sql general_user@localhost select * from f2025_w4111_examples.facultypublic;

In [ ]:
%sql general_user@localhost select * from f2025_w4111_examples.studentpublic;

I can allow ```general_user``` to create ```faculty``` and ```students``` without understanding the physical implementation. That is, ```general_user``` does not:
- Need ```insert``` priveleges.
- Does not need to understand that he/she has to ```insert``` into two tables.

In [ ]:
%%sql root@localhost

GRANT EXECUTE
    ON PROCEDURE f2025_w4111_examples.create_faculty
    TO 'general_user'@'%';

In [ ]:
%%sql general_user@localhost

use f2025_w4111_examples;

call create_faculty(
        "Douglas", "Fairbanks",
        "dfairbanks@cs.columbia.edu",
        "Professor",
        0,
        "COMS",
        @new_Uni
     );

In [ ]:
%sql general_user@localhost select * from Person

In [ ]:
%sql general_user@localhost select * from FacultyPublic

# Windows Functions

In [ ]:
%%sql

SELECT    country,    YEAR(orderDate) AS year,    ROUND(SUM(priceEach * quantityOrdered), 2) AS annual_revenue,    ROUND(        (SUM(priceEach * quantityOrdered) -        LAG(SUM(priceEach * quantityOrdered)) OVER (PARTITION BY country ORDER BY YEAR(orderDate))) /        LAG(SUM(priceEach * quantityOrdered)) OVER (PARTITION BY country ORDER BY YEAR(orderDate)) * 100, 2    ) AS yoy_growth_percentageFROM    customers    JOIN orders ON customers.customerNumber = orders.customerNumber    JOIN orderdetails ON orders.orderNumber = orderdetails.orderNumberGROUP BY    country, yearORDER BY    country, year;
